upstash_redis/asyncio/client.py

```diff
@@ -126,14 +126,132 @@ async def execute(self, command: List) -> RESTResultT:
                 command=command,
             )
 
-        main_command = command[0]
-        if len(command) > 1 and main_command == "SCRIPT":
-            main_command = f"{main_command} {command[1]}"
+        return cast_response(command, res)
 
-        if main_command in FORMATTERS:
-            return FORMATTERS[main_command](res, command)
+    def pipeline(self):
+        """
+        Create a pipeline to send commands in batches
+        """
+        return AsyncPipeline(
+            url=self._url,
+            token=self._token,
+            rest_encoding=self._rest_encoding,
+            rest_retries=self._rest_retries,
+            rest_retry_interval=self._rest_retry_interval,
+            allow_telemetry=self._allow_telemetry,
+            multi_exec="pipeline"
+        )
+
+    def multi(self):
+        """
+        Create a pipeline to send commands in batches as a transaction
+        """
+        return AsyncPipeline(
+            url=self._url,
+            token=self._token,
+            rest_encoding=self._rest_encoding,
+            rest_retries=self._rest_retries,
+            rest_retry_interval=self._rest_retry_interval,
+            allow_telemetry=self._allow_telemetry,
+            multi_exec="multi-exec"
+        )
+
+
+class AsyncPipeline(Redis):
+
+    def __init__(
+        self,
+        url: str,
+        token: str,
+        rest_encoding: Optional[Literal["base64"]] = "base64",
+        rest_retries: int = 1,
+        rest_retry_interval: float = 3,  # Seconds.
+        allow_telemetry: bool = True,
+        multi_exec: Literal["multi-exec", "pipeline"] = "pipeline"
+    ):
+        """
+        Creates a new blocking Redis client.
+
+        :param url: UPSTASH_REDIS_REST_URL in the console
+        :param token: UPSTASH_REDIS_REST_TOKEN in the console
+        :param rest_encoding: the encoding that can be used by the REST API to parse the response before sending it
+        :param rest_retries: how many times an HTTP request will be retried if it fails
+        :param rest_retry_interval: how many seconds will be waited between each retry
+        :param allow_telemetry: whether anonymous telemetry can be collected
+        :param miltiexec: Whether multi execution (transaction) or pipelining is to be used
+        """
+        super().__init__(
+            url=url,
+            token=token,
+            rest_encoding=rest_encoding,
+            rest_retries=rest_retries,
+            rest_retry_interval=rest_retry_interval,
+            allow_telemetry=allow_telemetry,
+        )
+        
+        self._command_stack: List[List[str]] = []
+        self._multi_exec = multi_exec
+
+    def execute(self, command: List) -> None: # type: ignore[override]
```

Comment:
Since we extend the Redis here, the IDEs pick up the command definitions from there and showing information that is not correct. Do you think we should do something about it? Also, retuning the self here would be a good idea, so someone can do `pipe = redis.pipeline(); pipe.incr("a").incr("b").exec()`

Reply:
mypy does not complain with type definitions here, as I think it does not properly work with the .pyi files we have that are separate from the code, but other type checkers might complain here (i haven't tested but pyright is another popular type checker, we can test what it does here)

Reply:
I fixed the return types and even our tests don't pass the mypy check because [these methods are not being awaited even though they are async](https://github.com/upstash/redis-python/pull/47/files#diff-a50b0906300b7e49eca6f5b4d8a7b44796afe8f26dd4726cb81a96185b27118cR16).

In the typescript client, the pipeline and redis methods are defined seperately for this reason ([redis](https://github.com/upstash/upstash-redis/blob/main/pkg/redis.ts#L589), [pipeline](https://github.com/upstash/upstash-redis/blob/main/pkg/pipeline.ts#L504)). I would prefer to avoid this but we may need to do something similar ultimately

Reply:
Maybe we can do this only for the type definitions, via a file similar to commands.pyi for pipeline. 

I think that test problem would also be solved with this, if we specialize type definitions for pipeline, and make it return None, or Pipeline object itself(if you want to make methods chainable)

Reply:
I fixed the errors by defining a Pipeline class in commands.pyi. I also made it possible to chain commands, see the tests for examples

only thing which looks wrong in the pipelines is the docstrings, but for that we would need to write out the whole class again. I think current state is alright

Also we decided to handle auto pipeline later ([linear](https://linear.app/upstash/issue/DX-907/redis-python-sdk-auto-pipeline-support)), so we can go ahead with this pr!